### PR TITLE
New rule S3416: Loggers should be named for their enclosing types

### LINKED ISSUE #8844: New rule S3416: Loggers should be named for their enclosing types
See https://github.com/SonarSource/rspec/pull/2528

### PR DESCRIPTION
**Pull request by gregory-paidis-sonarsource**
Fixes #8844

### CODE REVIEW

File: analyzers/tests/SonarAnalyzer.Test/TestCases/LoggersShouldBeNamedForEnclosingType.cs [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+
+        factory.CreateLogger(typeof(EnclosingType).FullName);           // Compliant
+        factory.CreateLogger(typeof(AnotherType).FullName);             // Noncompliant
+    }
+
+    void Types(ILoggerFactory factory)
+    {
+        // runtime
+        factory.CreateLogger(GetType());                                // Compliant
+        factory.CreateLogger(new AnotherType().GetType());              // Compliant FN, we do not parse this
```
Comment (zsolt-kolbay-sonarsource) [inline analyzers/tests/SonarAnalyzer.Test/TestCases/LoggersShouldBeNamedForEnclosingType.cs:32]:
```suggestion
        factory.CreateLogger(new AnotherType().GetType());              // FN, we do not parse this
```

File: analyzers/tests/SonarAnalyzer.Test/TestCases/LoggersShouldBeNamedForEnclosingType.cs [hunk 2]
```diff
@@ -0,0 +1,161 @@
+﻿using Microsoft.Extensions.Logging;
+using System;
+
+class AnotherType : EnclosingType { }
+
+class EnclosingType
```
Comment (zsolt-kolbay-sonarsource) [inline analyzers/tests/SonarAnalyzer.Test/TestCases/LoggersShouldBeNamedForEnclosingType.cs:6]: Add test case for the following:
- aliased type: `using AliasedFactory = Microsoft.Extensions.Logging.ILoggerFactory;`
- `using static` (e.g. for the `LogManager` in `NLog`)
- constant string value: private const string ConstLoggerName = "AnotherType"; (you can leave it as FN or implement it by retrieving the `ConstantValue` from `IFieldSymbol`)

**Review by zsolt-kolbay-sonarsource — APPROVED** [review]
LGTM! Left some polishing comments.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/LoggersShouldBeNamedForEnclosingType.cs [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+    }
+
+    private static SyntaxNode ExtractTypeOfName(MemberAccessExpressionSyntax memberAccess) =>
+        memberAccess.Expression is TypeOfExpressionSyntax typeOf
+        && memberAccess.GetName() is "Name" or "FullName" or "AssemblyQualifiedName"
+            ? typeOf.Type
+            : null;
+
+    private static SyntaxNode ExtractNameOf(InvocationExpressionSyntax invocation) =>
+        invocation.NameIs("nameof") && invocation.ArgumentList?.Arguments.Count == 1
```
[the remark this answers is not in the document]
Comment (gregory-paidis-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/LoggersShouldBeNamedForEnclosingType.cs:132]: I have no idea, I am just being defensive.
I am 99% sure that there is a case where it is null in `VB.NET`, but I do not know about C#.
Do you think we should remove it?

**Pull request merged**